File: ios/GitaVani/GitaVani/Services/GitaDataService.swift
import Foundation

@Observable
class GitaDataService {
    private(set) var chapters: [Chapter] = []
    private(set) var verses: [Verse] = []
    private(set) var isLoaded = false

    init() {
        loadData()
    }

    private func loadData() {
        guard let url = Bundle.main.url(forResource: "gita_data", withExtension: "json") else {
            print("GitaDataService: gita_data.json not found in bundle")
            return
        }

        do {
            let data = try Data(contentsOf: url)
            let decoder = JSONDecoder()
            decoder.keyDecodingStrategy = .convertFromSnakeCase
            let gitaData = try decoder.decode(GitaData.self, from: data)
            chapters = gitaData.chapters
            verses = gitaData.verses
            isLoaded = true
            print("GitaDataService: Loaded \(chapters.count) chapters, \(verses.count) verses")
        } catch {
            print("GitaDataService: Failed to load data - \(error)")
        }
    }

    func verses(forChapter chapterNumber: Int) -> [Verse] {
        verses.filter { $0.chapter == chapterNumber }
    }

    func verse(id: String) -> Verse? {
        verses.first { $0.id == id }
    }

    func chapter(number: Int) -> Chapter? {
        chapters.first { $0.chapterNumber == number }
    }
}

File: android/GitaVani/app/src/main/java/com/nikhilsi/gitavani/data/GitaDataService.kt
package com.nikhilsi.gitavani.data

import android.content.Context
import android.util.Log
import com.nikhilsi.gitavani.model.Chapter
import com.nikhilsi.gitavani.model.GitaData
import com.nikhilsi.gitavani.model.Verse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

class GitaDataService {
    var chapters: List<Chapter> = emptyList()
        private set
    var verses: List<Verse> = emptyList()
        private set
    var isLoaded: Boolean = false
        private set
    var loadError: String? = null
        private set

    private var verseById: Map<String, Verse> = emptyMap()
    private var versesByChapter: Map<Int, List<Verse>> = emptyMap()
    private var chapterByNumber: Map<Int, Chapter> = emptyMap()

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun loadData(context: Context) {
        withContext(Dispatchers.IO) {
            try {
                val jsonString = context.assets.open("gita_data.json")
                    .bufferedReader()
                    .use { it.readText() }

                val data = json.decodeFromString<GitaData>(jsonString)

                chapters = data.chapters
                verses = data.verses

                verseById = verses.associateBy { it.id }
                versesByChapter = verses.groupBy { it.chapter }
                chapterByNumber = chapters.associateBy { it.chapterNumber }

                isLoaded = true
                Log.d("GitaDataService", "Loaded ${chapters.size} chapters, ${verses.size} verses")
            } catch (e: Exception) {
                loadError = "Failed to load Gita data: ${e.message}"
                Log.e("GitaDataService", loadError!!, e)
            }
        }
    }

    fun versesForChapter(chapterNumber: Int): List<Verse> =
        versesByChapter[chapterNumber] ?: emptyList()

    fun verse(id: String): Verse? = verseById[id]

    fun chapter(number: Int): Chapter? = chapterByNumber[number]
}
